Cache Item computation

cdmihai opened the pull request:
On the test repo (solution with 72 VCX projects) this reduces evaluation by ~120ms. The improvement should get significantly better the more cross item references there are.

Items can cross-reference each other during evaluation via:
- item references (`Include="@(a)"`)
- conditions on item groups, items, and metadata (`Condition="@(a->Count()) != 0"`)
- metadata item references (`<m>@(a)</m>`)

In terms of memory, the cache pressure isn't that large because many other allocations are prevented by not doing the repeated computations, so the more cross item references there are, the more the cache pays for itself

rainersigwald commented:
I am looking at this but I have passed the part of the day where I can reason about the correctness of the lookahead remove operations. Tomorrow!

cdmihai commented:
@dsplaisted

cdmihai commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.LazyItemOperation.cs:
```diff
@@ -76,6 +61,22 @@ protected virtual ICollection<I> SelectItems(ImmutableList<ItemData>.Builder lis
 
             protected virtual void SaveItems(ICollection<I> items, ImmutableList<ItemData>.Builder listBuilder) { }
 
+            private IList<I> GetReferencedItems(string itemType, ImmutableHashSet<string> globsToIgnore)
```
Must have been some manual refactoring left-over. I'll undo the move.

cdmihai commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs:
```diff
@@ -133,34 +133,97 @@ public ItemData(I item, int elementOrder, bool conditionResult)
             public bool ConditionResult { get { return _conditionResult; } }
         }
 
+        private class OperationWithCache : IItemOperation
```
I like MemoizedOperation. It's just that it's not self-memoizing and needs an external caller to tell it when to cache.

cdmihai commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs:
```diff
@@ -133,34 +133,97 @@ public ItemData(I item, int elementOrder, bool conditionResult)
+                //   - has all operations uncached (nobody needed their results until now). In this case we have to compute the head plus the entire tail,
+                // but we don't need to cache each node of the tail because no external caller wanted their results
+                //   - has some operations cached (somebody needed their result at some point). In this case we have to compute the head plus the part of the 
+                // tail up to the first cached operation.
```
Yes, that would be the worst case. It happens if you have this:

```xml
1: <C Include="a"/>
2: <ForcesCacheOnC Include="@(C)"/>
3: <C Include="b"/>
4: <ForcesCacheOnC Include="@(C)"/>
5: <C Include="c"/>
6: <ForcesCacheOnC Include="@(C)"/>
```

However, the caches will share structural state. So the cache at operation `3` will share common state with the previously made cache at operation `1` (shared state is the collection with item 'a'). This is done via the immutable list. Also, when `GetItems` is called on operation `3`, it won't redo operation `1` but use its cache.

So no computation gets ever redone, but we will have a lot of `OperationWithCache` decorators if the above pattern is very common.

We could do other fancy things like reference counting (assuming an external caller calls `GetItems` only once, then we could decrement some CalleeNumber on each call per operation, and then `null` the cache for that operation when it has satisfied all of it's external callers). But I don't think it's worth making it more complex at this point until we get an outside ask about it :)

rainersigwald replied:
> I don't think it's worth making it more complex at this point

💯

rainersigwald commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.IncludeOperation.cs:
```diff
@@ -111,14 +127,27 @@ protected override ICollection<I> SelectItems(ImmutableList<ItemData>.Builder li
+                foreach (var item in items)
+                {
+                    listBuilder.Add(new ItemData(item, _elementOrder, _conditionResult));
+                }
```
Ahh, because this method is called so often! That totally makes sense, thanks.

cdmihai commented:
@dsplaisted That is indeed a real issue :(

It is curious though that the test I had written exactly for this scenario did not fail: https://github.com/Microsoft/msbuild/blob/xplat/src/XMakeBuildEngine/UnitTests/Evaluation/ItemEvaluation_Tests.cs#L91

I think it does not fail because the [Dictionary<string, LazyItemList>](https://github.com/Microsoft/msbuild/blob/xplat/src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs#L51) the lazy evaluator is using gets some happy key ordering which hides this issue. I'll play around with the test and try to get it failing.

The fix for the actual issue would be for Update to delete the old items and create new ones.

cdmihai commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.IncludeOperation.cs:
```diff
@@ -58,6 +58,19 @@ protected override ICollection<I> SelectItems(ImmutableList<ItemData>.Builder li
                     }
                 }
 
+                var excludePatternsForGlobs = new Lazy<ISet<string>>(() =>
```
Why would that be an improvement? The lazy setting before the first access vs using the Lazy seem to be doing the same thing (except creating that extra object for Lazy)

dsplaisted replied:
Yeah it avoids extra object creation for the lazy object itself and possibly the closure used for the creation delegate.

cdmihai commented:
- updated the caching code to ensure that each operation is applied at most once. Before, if a list had two item references to it (from other itemspecs or from conditions), and if the newer reference executed first, then the older reference would have re-computed the shared state between the two operations.
- redid the Update operation so it clones items before it mutates them

cdmihai commented:
@dotnet-bot test this please

dsplaisted commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs:
```diff
@@ -136,28 +137,68 @@ public ItemData(I item, int elementOrder, bool conditionResult)
             public void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
             {
+#if DEBUG
+                Debug.Assert(!_applyCalled, "Apply should only be called once. Otherwise caching is not working");
```
It should be valid for `Apply` to be called multiple times if the `globsToIgnore` is different each time.

dsplaisted commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs:
```diff
@@ -136,28 +137,68 @@ public ItemData(I item, int elementOrder, bool conditionResult)
+            /// <summary>
+            /// Somebody is referencing this operation
+            /// </summary>
+            public void AddReference()
```
It might be better to find a different name.  `AddReference` sounds to me like this is doing ref counting.

cdmihai commented:
@dotnet-bot test this please

dsplaisted commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs:
```diff
@@ -174,12 +215,19 @@ public LazyItemList(LazyItemList previous, LazyItemOperation operation)
+                // intermediary operations that would be requested by those callers.
+                //   - Callers of GetItems that cannot announce ahead of time. This includes item referencing conditions on
+                // Item Groups and Item Elements. However, those conditions are performed eagerly outside of the LazyItemEvaluator, so they will run before
+                // any item referencing operations from inside the LazyItemEvaluator. This 
```
Dangling "This"

dsplaisted commented on src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs:
```diff
@@ -117,50 +118,171 @@ private static string GetCurrentDirectoryForConditionEvaluation(ProjectElement e
+            public void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
+            {
+#if DEBUG
+                Debug.Assert(!_applyCalled, "Apply should only be called once. Otherwise caching is not working");
```
It should be valid for this method to be called multiple times if `globsToIgnore` is different each time.

(I left this comment on an outdated version of the file, so repeating it here in case that doesn't show up).

cdmihai replied:
Ah, forgot about that. The assert condition should be if the passed in `globsToIgnore ` is already cached. Also have to write a unit test to make this faulty assert fail :)

cdmihai commented:
@dsplaisted 

Fixed the assert to: number of calls should be equal to cache size.
Also fixed a hidden bug that I introduced: the memoized operation no longer cast to RemoveOperation, so no globs flowed to previous operations. Hard to add an assert for this. One way to add an assert is to change Remove so that it flows all of its values (literals, item references, and globs) not just globs, and Remove is skipped altogether. With this change, we can just throw in Remove's apply :)
